File: java/anagram/src/main/java/Anagram.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Anagram {
    private String word;

    public Anagram(String word) {
        this.word = word;
    }

    public List<String> match(List<String> strings) {
        List<String> result = new ArrayList<>();

        for (String candidate : strings) {
            if (haveDifferentSize(word, candidate))
                continue;

            if (isAnagram(word, candidate)) {
                result.add(candidate);
            }
        }
        return result;
    }

    private boolean isAnagram(String word, String candidate) {
        word = word.toLowerCase();
        candidate = candidate.toLowerCase();

        return !word.equals(candidate) && convert(word).equals(convert(candidate));
    }

    private boolean haveDifferentSize(String word, String candidate) {
        return word.length() != candidate.length();
    }

    private String convert(String s) {
        char[] ar = s.toCharArray();
        Arrays.sort(ar);
        return String.valueOf(ar);
    }
}

File: go/anagram/anagram.go
package anagram

import (
	"sort"
	"strings"
)

//Detect returns the list of anagram of the input "word" contained in "anagrams"
func Detect(word string, anagrams []string) []string {
	word = strings.ToLower(word)
	result := []string{}

	for _, candidate := range anagrams {
		candidate = strings.ToLower(candidate)

		if haveDifferentSize(word, candidate) {
			continue
		}

		if isAnagram(word, candidate) {
			result = append(result, candidate)
		}

	}
	return result
}

func haveDifferentSize(word, candidate string) bool {
	return len(candidate) != len(word)
}

func isAnagram(word, candidate string) bool {
	return word != candidate && convert(word) == convert(candidate)
}

func convert(s string) string {
	sNoSpace := strings.Split(s, "")
	sort.Strings(sNoSpace)
	return strings.Join(sNoSpace, "")
}
